File: Leetcode/max_profit.py
# LeetCode task: solution -> https://leetcode.com/problems/best-time-to-buy-and-sell-stock/description/?envType=study-plan-v2&envId=top-interview-150
# 🎯 150 task preparation to the interview 😎

from typing import List


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        """
        Calculates the maximum profit that can be achieved by buying and selling a stock.

        Args:
            prices (List[int]): The list of stock prices.

        Returns:
            int: The maximum profit that can be achieved.

        """

        # Check if there are at least 2 prices in the list
        if len(prices) < 2:
            return 0

        # Initialize variables
        min_price = prices[0]  # Track the minimum price seen so far
        max_profit = 0  # Track the maximum profit seen so far

        # Iterate over the prices list
        for i in range(1, len(prices)):
            if prices[i] < min_price:
                # Update the minimum price if a smaller price is encountered
                min_price = prices[i]
            else:
                # Calculate the potential profit if sold at the current price
                profit = prices[i] - min_price
                # Update the maximum profit if necessary
                max_profit = max(max_profit, profit)

        return max_profit


prices = [7, 1, 5, 3, 6, 4]

# Create an instance of the Solution class
solution = Solution()

# Call the maxProfit method to calculate the maximum profit
max_profit = solution.maxProfit(prices)

# Print the result
print("Maximum Profit:", max_profit)

File: CSharp/max_profit.cs
using System;

public class Solution
{
    /// <summary>
    /// Calculates the maximum profit that can be achieved from buying and selling stock.
    /// The function assumes that you can only complete one transaction (i.e., buy one and sell one share of the stock).
    /// </summary>
    /// <param name="prices">An array of integers representing the stock prices on different days.</param>
    /// <returns>The maximum profit that can be achieved. If no profit can be made, returns 0.</returns>
    public int MaxProfit(int[] prices)
    {
        // Check if there are at least 2 prices in the array
        if (prices.Length < 2)
        {
            return 0; // Not enough prices to make a transaction
        }

        int minPrice = prices[0]; // Initialize the minimum price to the first day's price
        int maxProfit = 0; // Initialize the maximum profit to 0

        // Iterate over the prices array starting from the second price
        for (int i = 1; i < prices.Length; i++)
        {
            // Update the minimum price if a smaller price is encountered
            if (prices[i] < minPrice)
            {
                minPrice = prices[i];
            }
            else
            {
                // Calculate the potential profit if sold at the current price
                int profit = prices[i] - minPrice;
                // Update the maximum profit if the current profit is greater
                maxProfit = Math.Max(maxProfit, profit);
            }
        }

        return maxProfit; // Return the maximum profit found
    }
}

class Program
{
    static void Main(string[] args)
    {
        // Example stock prices for testing
        int[] prices = { 7, 1, 5, 3, 6, 4 };

        // Create an instance of the Solution class
        Solution solution = new Solution();

        // Call the MaxProfit method to calculate the maximum profit
        int maxProfit = solution.MaxProfit(prices);

        // Print the result
        Console.WriteLine("Maximum Profit: " + maxProfit);
    }
}
